src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java

```diff
@@ -60,6 +58,20 @@ class ChannelInputStream extends InputStream {
         this.ch = ch;
     }
 
+    /*
+     * Reads at most Streams.DEFAULT_BUFFER_SIZE bytes into the given buffer.
+     * If the channel is selectable then it must be configured blocking.
+     */
+    private int readBytes(ByteBuffer bb) throws IOException {
+        int pos = bb.position();
+        int rem = bb.limit() - pos;
+        if (rem <= 0)
+            return 0;
+
+        bb.limit(pos + Integer.min(rem, Streams.DEFAULT_BUFFER_SIZE));
```

Comment:
I assume we would end up with a MAX_BUFFER_SIZE that is something like 128k, like we ended up in NioSocketImpl for the same reason. In order words, it's different to the chunk size that is DEFAULT_BUFFER_SIZE today.

Reply:
FWIW, getting back to the reason that this was created in the first place, namely that it is possible to limit `MaxDirectBufferSize`, ideally this limit could be dependent on how much is actually available in the system. `Bits` doesn't currently have any way to expose that. In the scenario that I was dealing with, I had 28M of direct buffer, but the system was trying to allocate (roughly) 62M, and that didn't work. So limiting to 128K would still work perfectly in that scenario.